pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -546,4 +561,12 @@ std::unique_ptr<mlir::Pass> createStencilPass() {
   return std::make_unique<StencilPass>();
 }
 
+std::unique_ptr<mlir::Pass> createStencilPassWithCoster(
```

Comment:
Why not just change the constructor to the original createStencilPass?  Does having one without a coster make any sense.

Reply:
Maybe someone can point how to register a pass with constructor function with parameters. I have not seen anything like this in the code and looking at the PassRegistry.h, the only way to register a Pass is with no argument function.
Here is some code...
template <typename ConcretePass> struct PassRegistration {
  PassRegistration(StringRef arg, StringRef description,
                   const PassAllocatorFunction &constructor) {
    registerPass(arg, description, PassID::getID<ConcretePass>(), constructor);
  }
using PassAllocatorFunction = std::function<std::unique_ptr<Pass>()>;
Basically, the registration code fails to compile with the following erroir...
pmlc/dialect/pxa/transforms/registration.cc:35:5: error: no matching constructor for initialization of 'mlir::PassRegistration<StencilPass>'
    stencil_pass("affine-stencil",
    ^            ~~~~~~~~~~~~~~~~~
external/llvm-project/mlir/include/mlir/Pass/PassRegistry.h:139:3: note: candidate constructor not viable: no known conversion from 'std::unique_ptr<mlir::Pass> (std::function<std::pair<double, unsigned int> (const unsigned int *, const unsigned int)>)' to 'const mlir::PassAllocatorFunction' (aka 'const function<std::unique_ptr<Pass> ()>') for 3rd argument
  PassRegistration(StringRef arg, StringRef description,
  ^
external/llvm-project/mlir/include/mlir/Pass/PassRegistry.h:144:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
  PassRegistration(StringRef arg, StringRef description)

There is currently a function in the code that is the "default" coster - return 1.0 for each tile. And then the user can privide it's own. If any of you have a snipped/doc describing how to register passes with parameters in the registration process, I'd really appreciate a pointer. I just could not find any in our/llvm code bases.

Reply:
Keep in mind that PassRegistration is only needed for the `pmlc-opt` tool, which is really only intended for testing. So what I would do is have a named PassRegistration that builds a specific stencil pass with a specific coster. Call it the  `-xsmm-stencil-pass` or something. I don't think we need to worry about making the PassRegistration generic itself.

Reply:
Also, I'd register this specific pass over in the `pmlc/target/x86` area, since it's X86 specific and will need access to the heatmap, which is also x86 specific.

Reply:
@flaub Yes! That is exactly what I figured and implemented it this way. The creasteStencilPass function does exactly that... It register the pass and uses a default coster function (the default one). The createStencilPassWithCoster is not registered and is intended to be instantiated by the pm.addPass(createStencilPassWithCoster(function)) from the pipeline directly the same way ConstantTypesPass in the tile layer is instantiated (pm.addPass(createConstantTypesPass(floatx, intx));)

Reply:
Moving the registration out of the stencil pass and in the x86 is an interesting idea but it could mean that the whole pass needs to be moved. We are testing its logic now with a "default" coster function. And to specify another is not feasible with no ability to register it - again the tests for the pass are running with opt, so it needs to be registered.